decred/decred/dcr/txscript.py

```diff
@@ -3509,7 +3510,213 @@ def estimateSerializeSizeFromScriptSizes(inputSizes, outputSizes, changeScriptSi
+        votes.  Consequently, this does not accept the number of votes.  However, it
+        is important to note that blocks that do not receive the minimum required
+        number of votes for a block to be valid by consensus won't actually produce
+        any vote subsidy either since they are invalid.
+
+        This function is safe for concurrent access.
+        """
+        # Votes have no subsidy prior to the point voting begins.  The minus one
+        # accounts for the fact that vote subsidy are, unfortunately, based on the
+        # height that is being voted on as opposed to the block in which they are
+        # included.
+        if height < self.netParams.StakeValidationHeight - 1:
+            return 0
+
+        # Calculate the full block subsidy and reduce it according to the stake
+        # proportion.  Then divide it by the number of votes per block to arrive
+        # at the amount per vote.
+        subsidy = self.calcBlockSubsidy(height)
+        proportions = self.totalProportions
+        subsidy *= self.netParams.StakeRewardProportion
+        subsidy = subsidy // (proportions * self.netParams.TicketsPerBlock)
+
+        return subsidy
+
+    def calcTreasurySubsidy(self, height, voters):
+        """
+        calcTreasurySubsidy returns the subsidy required to go to the treasury for
+        a block.  It is calculated as a proportion of the total subsidy and further
+        reduced proportionally depending on the number of votes once the height at
+        which voting begins has been reached.
+
+        Note that passing a number of voters fewer than the minimum required for a
+        block to be valid by consensus along with a height greater than or equal to
+        the height at which voting begins will return zero.
+
+        This function is safe for concurrent access.
+        """
+        # The first two blocks have special subsidy rules.
+        if height <= 1:
+            return 0
+
+        # The subsidy is zero if there are not enough voters once voting begins.  A
+        # block without enough voters will fail to validate anyway.
+        stakeValidationHeight = self.netParams.StakeValidationHeight
+        if height >= stakeValidationHeight and voters < self.minVotesRequired:
+            return 0
+
+        # Calculate the full block subsidy and reduce it according to the treasury
+        # proportion.
+        subsidy = self.calcBlockSubsidy(height)
+        subsidy *= self.netParams.BlockTaxProportion
+        subsidy = subsidy // self.totalProportions
+
+        # Ignore any potential subsidy reductions due to the number of votes prior
+        # to the point voting begins.
+        if height < stakeValidationHeight:
+            return subsidy
+
+        # Adjust for the number of voters.
+        return (voters * subsidy) // self.netParams.TicketsPerBlock
+
+
+def stakePoolTicketFee(stakeDiff, relayFee, height, poolFee, netParams):
```

Comment:
We should still allow the cache to be passed in. Creating a new cache every time defeats the purpose. Could be an optional argument though.

Reply:
Ah, right, silly me. In that case I'd put the class in `dcrdata.py` rather than here.

Reply:
Even though `DcrdataBlockchain` is the only user currently, the only place the cache is actually used is here in txscript. I imagine the soon-to-be `Blockchain` based on a local dcrd node will also use one. Let's leave the class in **txscript**.